microcosm_eventsource/func.py

```diff
@@ -0,0 +1,67 @@
+"""
+Custom database functions.
+
+"""
+from sqlalchemy.sql.expression import FunctionElement
+from sqlalchemy.ext.compiler import compiles
+
+
+class last(FunctionElement):
+    """
+    Define a SQLAlchemy function that maps to a postgres function to select the last non-null value in window.
+
+    """
+    name = "last"
+
+    @staticmethod
+    def create(connection):
+        """
+        Create the postgres functions for this implementation.
+
+        XXX Not currently called in any automatic way.
+
+        """
+        connection.execute("""
+          CREATE OR REPLACE FUNCTION last_agg_sfunc (state anyelement, value anyelement)
+                 RETURNS anyelement
+                 LANGUAGE SQL
+                 IMMUTABLE
+          AS $$
+            SELECT coalesce(value, state);
+          $$;
+        """)
+        connection.execute("""
+          CREATE AGGREGATE last_agg (anyelement) (
+                SFUNC = last_agg_sfunc,
+                STYPE = anyelement
+          );
+        """)
+
+    @staticmethod
+    def drop(connection):
+        """
+        Drop the postgres functions for this implementation.
+
+
+        XXX Not currently called in any automatic way.
+        """
+        connection.execute("DROP AGGREGATE IF EXISTS last_agg (anyelement);")
+        connection.execute("DROP FUNCTION IF EXISTS last_agg_sfunc (state anyelement, value anyelement);")
+
+
+@compiles(last)
+def compile(element, compiler, **kwargs):
+    """
+    By default, `last` is not defined.
+
+    """
+    raise NotImplementedError("last is only defined for postgresql")
+
+
+@compiles(last, "postgresql")
+def compile_postgres(element, compiler, **kwargs):
+    """
+    Define `last` in terms of the custom functions.
+
+    """
+    return f"last_agg({compiler.process(element.clauses)})"
```

Comment:
In other words, invoke our `last_agg` function from L34, which in turn invokes `last_agg_sfunc` from L25.